pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java

```diff
@@ -398,8 +442,44 @@ private static HepProgram getTraitProgram() {
     }
 
     // apply RelDistribution trait to all nodes
+    if (useImplicitColocated) {
+      hepProgramBuilder.addRuleInstance(PinotImplicitTableHintRule.withPartitionTableFinder(tablePartitionTableFinder));
+    }
     hepProgramBuilder.addRuleInstance(PinotRelDistributionTraitRule.INSTANCE);
 
     return hepProgramBuilder.build();
   }
+
+  public static ImmutableQueryEnvironment.Config.Builder configBuilder() {
+    return ImmutableQueryEnvironment.Config.builder();
+  }
+
+  @Value.Immutable
```

Comment:
Not sure if we want to use auto-gen annotations. We removed `lombok` in the past

Reply:
This is not like Lombok. Lombok works post compilation, changing the bytecode to add the generated stuff. Contrary to that, this library works before compilation, generating source classes with java code that contains the logic included here. The difference is key. For example you can debug the generated code and the IDE doesn't need to do anything crazy to understand the generated code. It is similar to the way we generate the parser or protobuf code.

This is the way Calcite creates its own rules and it is how they recommend to create new rules in [their tutorial](https://calcite.apache.org/docs/tutorial.html#optimizing-queries-using-planner-rules).

Reply:
Understood that this is the recommended way when adding Calcite rules for consistency. My question is around whether we want to take this approach in Pinot code. This is `QueryEnvironment`, and I don't really see why we want to take the `Config` approach

Reply:
I think adding a `Config` object makes sense given the number of arguments in the constructor is increasing. Whether that `Config` object is created manually or using a tool (like this `immutables` library) is something we can discuss. Given we are already adding the library to follow Calcite's best practices, I don't think it is going to hurt to use the same library in other POJOs that are simple enough but verbose. `immutables` also has the nice property of being pretty expressive in terms of nullability and optional values, which is also great to do not miss to add the `@Nullable` annotations in these POJOs.

Anyway, given it generates normal Java code, it would be easy to remove the usage of `immutables` in the future if we don't like.

Reply:
Are you ok merging it using immutables or should I change the code to use a manually written POJO?

Reply:
I'm okay with it since it is easy to change